EngineEffectsDelay: introduce ring delay buffer

davidchocholaty opened the pull request:
This PR introduces the `RingDelayBuffer` and replaces the previous processing approach. Based on that, the implemented crossfade is replaced with `SampleUtil::linearCrossfadeBuffersOut`. Dependent tests are upgraded because `RingDelayBuffer` uses the fading-in, for the first written chunk of data, and the crossfading is used in stereo mode.

davidchocholaty committed:
SampleBuffer: add subspan function

This commit adds the SampleBuffer::subspan function as an additional
function to implement SampleBuffer::span. The two independent subspan
functions were implemented because otherwise it is necessary to as first
create std::span from SampleBuffer using SampleBuffer::span
and then create from created span subspan
using std::span<T,Extent>::subspan from the standard library.
The SampleBuffer's implementation avoids creating two spans
and allows to creation subspan directly from SampleBuffer.

davidchocholaty committed:
EngineEffectsDelay: introduce ring delay buffer

This commit introduces the RingDelayBuffer and replaces the previous
processing approach. Based on that, the implemented crossfade
is replaced with a function from SampleUtil, concretely
SampleUtil::linearCrossfadeBuffersOut. Dependent tests are upgraded
because RingDelayBuffer uses the fading-in, for the first written chunk
of data, and the crossfading is used in stereo mode.

Swiftb0y commented on src/engine/effects/engineeffectsdelay.h:
```diff
@@ -76,6 +78,6 @@ class EngineEffectsDelay final : public EngineObject {
-    SINT m_delayBufferWritePos;
-    CSAMPLE* m_pDelayBuffer;
+    mixxx::SampleBuffer m_currentDelayBuffer;
+    std::unique_ptr<RingDelayBuffer> m_pDelayBuffer;
```
I don't see any reason for allocating here. RingDelayBuffer already does internally. 
```suggestion
    RingDelayBuffer m_pDelayBuffer;
```

Swiftb0y commented on src/engine/effects/engineeffectsdelay.cpp:
```diff
@@ -1,78 +1,47 @@
+        // TODO(davidchocholaty) check the returned number of read samples
+        // Read the samples using the previous group delay samples.
+        m_pDelayBuffer->read(inOutSpan, m_prevDelaySamples);
+        // Read the samples using the current group delay samples.
+        m_pDelayBuffer->read(m_currentDelayBuffer.subspan(0, iBufferSize), m_currentDelaySamples);
+
+        SampleUtil::linearCrossfadeBuffersOut(
+                pInOut,
+                m_currentDelayBuffer.data(),
+                iBufferSize);
```
A shame that we need to copy here... iterators could solve that problem. 

The suggestion below is identical in functionality but is a little better self-documenting in my opinion.

```suggestion
        // Read the samples using the current group delay samples.
        std::span<CSAMPLE> tmpBufferView = m_currentDelayBuffer.span().first(inOutSpan.size());
        m_pDelayBuffer->read(tmpBufferView, m_currentDelaySamples);
        
        SampleUtil::linearCrossfadeBuffersOut(
                pInOut,
                tmpBufferView.data(),
                tmpBufferView.size());
```

Swiftb0y reviewed (changes requested):
a couple suggestions. looks very good already otherwise.

davidchocholaty commented on src/util/samplebuffer.h:
```diff
@@ -81,6 +81,23 @@ class SampleBuffer final {
         return mixxx::spanutil::spanFromPtrLen(m_data, m_size);
     }
 
+    std::span<CSAMPLE> subspan(const SINT offset, const SINT count) {
+        DEBUG_ASSERT((m_data != nullptr) || ((offset == 0) && (count == 0)));
+        DEBUG_ASSERT(0 <= offset);
+        DEBUG_ASSERT(0 <= count);
+        DEBUG_ASSERT(m_size > offset + count);
+
+        return mixxx::spanutil::spanFromPtrLen(m_data + offset, count);
+    }
+    std::span<const CSAMPLE> subspan(const SINT offset, const SINT count) const {
+        DEBUG_ASSERT((m_data != nullptr) || ((offset == 0) && (count == 0)));
+        DEBUG_ASSERT(0 <= offset);
+        DEBUG_ASSERT(0 <= count);
+        DEBUG_ASSERT(m_size > offset + count);
+
+        return mixxx::spanutil::spanFromPtrLen(m_data + offset, count);
+    }
+
```
Yeah, I thought about creating another span (subspan), from the span which we didn't need, so we have actually two spans with one unused. But as you describe the situation with inlining, it does not make much sense then.

davidchocholaty committed:
SampleBuffer: remove subspan function

The SampleBuffer::subspan are removed due to it makes more sense
to create std::span<T,Extent>::subspan on the span got
from SampleBuffer::span.

Swiftb0y commented:
you forgot to replace the place where you actually used `SampleBuffer::subspan`. Just change those and `amend` the commit.

davidchocholaty commented:
> you forgot to replace the place where you actually used `SampleBuffer::subspan`. Just change those and `amend` the commit.

Yeah, thank you. I figured it out now too. It should be fixed.

davidchocholaty committed:
EngineEffectsDelay: remove std::unique_ptr

The commit removes the std::unique_ptr for the delay buffer
and replaces it just with RingDelayBuffer, due to there is not
any reason for allocating and RingDelayBuffer already does it
internally.

Swiftb0y commented on src/engine/effects/engineeffectsdelay.cpp:
```diff
@@ -36,7 +36,9 @@ void EngineEffectsDelay::process(CSAMPLE* pInOut,
         // Read the samples using the current group delay samples.
-        m_pDelayBuffer->read(m_currentDelayBuffer.subspan(0, iBufferSize), m_currentDelaySamples);
+        m_pDelayBuffer->read(
+                m_currentDelayBuffer.span().subspan(0, iBufferSize),
```
```suggestion
                m_currentDelayBuffer.span().first(iBufferSize),
```

davidchocholaty replied:
Should be done with a fix for your previous proposal for the same part of the code. I mean the version using the `tmpBufferView `variable and primarily `inOutSpan.size()` for the size. IMO a more solid solution preventing errors in case of code changes.

davidchocholaty committed:
EngineEffectsDelay: use first instead of subspan

The commit replaces the std::span<T,Extent>::subspan with parameters
subspan(0, iBufferSize) with std::span<T,Extent>::first with just one
parameter, that means first(iBufferSize). This version is much more
readable and self-documenting with the same functionality.

Swiftb0y commented on src/test/engineeffectsdelay_test.cpp:
```diff
@@ -86,7 +86,7 @@ TEST_F(EngineEffectsDelayTest, DelayGreaterThanDelayBufferSize) {
-    const CSAMPLE expectedResult[] = {-100.0, 75.0, -49.5, 24.75};
+    const CSAMPLE expectedResult[] = {-50.0, 50.0, -49.5, 49.5};
 
     mixxx::SampleBuffer pInOut(numSamples);
```
I just noticed another flaw in these tests (sorry for widening the scope of the PR). 
the `p` prefix of variables is only used for variables holding a pointer. But samplebuffer is not a pointer, so we should fix the name accordingly.

davidchocholaty replied:
No problem. You're right, it does not make sense.

Swiftb0y commented on src/test/engineeffectsdelay_test.cpp:
```diff
@@ -105,9 +105,9 @@ TEST_F(EngineEffectsDelayTest, WholeBufferDelay) {
 
     const CSAMPLE inputBuffer[] = {-100.0, 100.0, -99.0, 99.0};
     const CSAMPLE zeroBuffer[] = {0.0, 0.0, 0.0, 0.0};
-    const CSAMPLE firstExpectedResult[] = {-100.0, 75.0, -49.5, 24.75};
+    const CSAMPLE firstExpectedResult[] = {-50.0, 50.0, -49.5, 49.5};
```
Having to change the data in tests to fix them is usually a sign that some observable behavior has changed. Can you explain why the change of the values is still correct?

davidchocholaty replied:
Yeah of course. I will start with the change in the `EngineEffectsDelay`.

1. In the version before, the ramping was used for samples chunk independent of the channel count, so the previous version sequence was [-100.0, 75.0, -49.5, 24.75]. Based on the change of use `SampleUtil::linearCrossfadeBuffersOut` the crossfade (in this case more ramping due to crossfade with zero "silence" samples) works in the I would call "stereo mode". So the new expected result is [-100.0, 100.0, -49.5, 49.5].

2. Due to the feature in the `RingDelayBuffer`, that the first chunk of written samples is faded-in due to avoid a "crackling" sound when crossing the border from zero "silence" samples to real data, the expected result should change based on this feature. So, the final result is [-50.0, 50.0, -49.5, 49.5].

Swiftb0y replied:
Ah yes, that makes total sense. Thank you for the explanation.

davidchocholaty committed:
EngineEffectsDelay: remove the p prefix

The commit removes the p prefix for the SampleBuffer's variable
because it does not hold a pointer.

davidchocholaty commented:
The failing test for macOS CI:

```
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from EngineEffectsDelayTest
[ RUN      ] EngineEffectsDelayTest.DelayGreaterThanDelayBufferSize
../src/test/engineeffectsdelay_test.cpp:34: Failure
Expected equality of these values:
  buffer[i]
    Which is: 1.539751e-39
  referenceBuffer[i]
    Which is: 0
[  FAILED  ] EngineEffectsDelayTest.DelayGreaterThanDelayBufferSize (2 ms)
[----------] 1 test from EngineEffectsDelayTest (2 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (2 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] EngineEffectsDelayTest.DelayGreaterThanDelayBufferSize

 1 FAILED TEST
```

The problem should be fixed probably with the #10853.

davidchocholaty commented on src/test/engineeffectsdelay_test.cpp:
```diff
@@ -86,13 +86,17 @@ TEST_F(EngineEffectsDelayTest, DelayGreaterThanDelayBufferSize) {
+    // Read with the maximum possible delay (currently
+    // they are zero "silence" samples - the start
+    // of the ring delay buffer).
+    const CSAMPLE expectedResult[] = {0.0f, 0.0f, 0.0f, 0.0f};
```
Oh, thank you for the proposal. I will try it.

davidchocholaty commented:
Perfect, thank you for the information. :+1:

davidchocholaty committed:
EngineEffectsDelay: check returned values

This commit adds a few VERIFY_OR_DEBUG_ASSERTs to check the written
or read number of samples. If the check is false, based on the situation
it is returned from the EngineEffectsDelay::process or the operation
is processed under specific circumstances.

davidchocholaty commented:
To sum up the status of this PR. The test for testing the situation when the delay is greater than the buffer size was modified. IMHO it makes more sense to test the structure more strictly. So, the whole delay buffer is filled with a linearly increasing series and with that, it can be tested that the received data are exactly as expected. For the previous version, when the whole delay buffer was filled with zeros, there is no guarantee that the correct data has been read (the zero samples from the correct indexes). With that, it does not make much sense to me to delay this PR with the rounding problem for macOS CI, which is not for the audio use case essential.

Swiftb0y commented:
So I guess the failure on MacOS was a out-of-bounds buffer read which caused the garbage values? Not Good...

Swiftb0y commented on src/test/engineeffectsdelay_test.cpp:
```diff
@@ -82,16 +82,33 @@ TEST_F(EngineEffectsDelayTest, DelayGreaterThanDelayBufferSize) {
+    // was written into the delay buffer too. So, when it will be read
+    // with the maximum possible delay, the result is the buffer, which was
+    // filled into the delay buffer as the second.
+    const CSAMPLE expectedResult[] = {4, 5, 6, 7};
```
Can you simplify this somehow? Tests need to be simple enough to confirm their correctness by simply looking at them. In this case, I not only can't confirm its correctness, I even struggle to understand what this does in the first place.

acolombier commented:
Hey @davidchocholaty, are you still interested in completing this? AFAIS, this only needed a small adjustment on testing and was ready to go! I'll mark it as draft for now, but feel free to revive if you get a chance to address it, or close if you don't have interest in this anymore.

davidchocholaty commented:
Hi, thank you for pinging me. I will take a look at how relevant this feature is and whether it can still be used this way. Based on that, I will either complete it or close it.
